### PR TITLE
Mark structs as readonly where appropriate

### PR DESCRIPTION
**Pull request by sharwell**
⚠️ While these cases are all readonly today, it is a breaking semantic change to change a `readonly struct` by removing the `readonly` keyword. Any types for which you would like to reserve the right to make the type mutable in the future should be excluded from this change.

With that said, at this point I'm not sure you would actually need to make any changes like this.

### CODE REVIEW

**Comment by AArnott** [comment]
Thanks. Can you tell me what we gain by marking these readonly? Is it that calls into the struct avoids a struct copy?

**Comment by AArnott** [comment]
@sharwell, I'm interested in moving this PR forward. Can you help me understand the answer to my above question?

**Comment by sharwell** [comment]
> Is it that calls into the struct avoids a struct copy?

Yes, this is correct. It allows the compiler to further optimize the IL code where these types are used.

File: src/Microsoft.VisualStudio.Threading.Tests/AsyncReaderWriterLockTests.cs [hunk 1]
```diff
@@ -5013,18 +5013,17 @@ protected override Exception OnCriticalFailure(Exception ex)
                 throw new CriticalErrorException(ex);
             }
 
-            internal struct InternalLockHandle
+            internal readonly struct InternalLockHandle
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading.Tests/AsyncReaderWriterLockTests.cs]: Wow. You're even optimizing _test_ structs? Did you have an analyzer that did all this for you?

Comment (Therzok) [reply src/Microsoft.VisualStudio.Threading.Tests/AsyncReaderWriterLockTests.cs]: Looks like ErrorProne.NET https://blogs.msdn.microsoft.com/seteplia/2018/05/03/avoiding-struct-and-readonly-reference-performance-pitfalls-with-errorprone-net/

**Pull request merged**